include/swift/SIL/SILInstruction.h

```diff
@@ -87,6 +88,83 @@ SILInstructionKind getSILInstructionKind(StringRef InstName);
 /// Map SILInstructionKind to a corresponding SILInstruction name.
 StringRef getSILInstructionName(SILInstructionKind Kind);
 
+/// A formal SIL reference to a list of values, suitable for use as the result
+/// of a SILInstruction.
+///
+/// *NOTE* Most multiple value instructions will not have many results, so if we
+/// want we can cache up to 3 bytes in the lower bits of the value.
+///
+/// *NOTE* Most of this defined out of line further down in the file to work
+/// around forward declaration issues.
+class SILInstructionResultArray {
+  const uint8_t *Pointer;
+  unsigned Size;
+  unsigned Stride;
+
+public:
+  SILInstructionResultArray(const NonValueInstruction *NVI);
+  SILInstructionResultArray(const SingleValueInstruction *SVI);
+
+  SILValue operator[](size_t Index) const {
+    unsigned Offset = Stride * Index;
+    return SILValue(reinterpret_cast<const ValueBase *>(&Pointer[Offset]));
```

Comment:
Hmm.  We're assuming here that the ValueBase subobject is at offset 0 within the array elements, but I guess that works for now.

Reply:
Given that we want everything to be static except in the constructor, I think we pretty much have to assume the offset (and in the constructor compensate dynamically). Unless (I guess), you are suggesting that we store the ValueBase subobject offset?